README.md
# WebApp

Small web application that allows text file manipulation. User can:

- List the files stored in a fix directory
- Create new files
- Edit new files

The backend is built with Go and the front end with HTML go templates along with Bulma CSS.

### Running the app

To run the app:

- Clone repo
- cd to the repo directory
- Build the app:
```bash
    go build
```
- call the executable
```bash
    webApp
```
- In your browser, go to:
```bash
    http://localhost:8080/
```
Note that depending on your system, you may be asked to allow the application to accept connections.

### Testing the app

To run all the packages within the app:

- cd to the root of the repo
- Execute tests:
```bash
go test ./...
```

To run the tests for a particular package:
- cd to the directory containing the package
- Execute test:
```bash
go test
```

Finally, if you want more details, you can use the -json flag eg:
```bash
go test -json
``` 

### Documentation
In typical Go manner the documentation is extracted directly from the comments in the application.

#### CLI

To view the high level documentation in the cli, run:
```bash
go doc
```
To view the documentation for publicly exposed functions, run:
```bash
go doc <function name>
```
To view the documentation for private functions, run:
```bash
go doc -u <function name>
```
To view all the documentation except for private functions
```bash
go doc -all
```
To view all the documentation including private functions
```bash
go doc -u - all
```
### Docker

The application has been setup to be able to run using docker and docker-compose.

To build the docker image:
```bash
docker build -t gowebapp .
```

To run the app using the docker file:

```bash
docker run -p 8080:8080 --name gowebappcont gowebapp
```

For docker-compose, things got a bit more interesting since the host machine is running MacOS while the image is based on the golang linux image.
As such, wehn we mount the volume from the Mac to th elinux image, the application binary is not compatible with one of the two environments.
To get around this, there are two options:

- If you do not care to share the code between the host and the guest OS, then simply comment the lines in the docker-compose fiie and you can run normally using ```docker-compose up``` and ```docker-compose down```
```bash
   volumes:
      - .:/go/src/goWebApp
```

- If you want to be able to share the code, then use the commands below instead of docker-compose up and docker-compose down. The commands below will create/delete linux binaries of the application (goWebAppLin) so that they can be run in the container normally.
```bash
./mycomposeup
```
```bash
./mycomposedown
```

#### Hot reload

The app is set up to automatically hot reload when running via docker-compose. It uses ```CompileDaemon``` to check for changes to .go files.

If you would like to have the same effect when running via the docker file directly (docker run), then uncomment the ```entrypoint``` and comment out the ```cmd``` in the ```Dockerfile```

#### Multistage build

The ```Dockerfilemulti``` file creates a minimal container from the app and the Alpine linux base image. The file has two stages:

1. Creates a docker image based on the golang official base image. This is basically the same image created by the original Dockerfile. The only difference is that this image will also contain a binary called ```goWebAppLinAlp``` which is created to be able to run 
our app in a linux Alpine environment.
2. Creates a linux Alpine image that can run the application

Size biggest difference between the golang image and the Alpine image. While the golang image is 858MB, the Alpine version is only 16MB (about **98%** size reduction).

To create an image using the ```Dockerfilemulti```  multistage build, run:
```bash
docker build -t gowebappalpine -f Dockerfilemulti .      

```
To run the above image:
```bash
docker run -p 8080:8080 --name gowebappalpinecont gowebappalpine  
```
**Note**: We could technically delete the original Docker file and just use the multistage version. To create an image similar to the golang image curently created by the original Docker file using the multistage build, we could just run:
```bash
docker build --target Dev  -t gowebappfs -f Dockerfilemulti .
```

### Kubernetes

The application is kubernetes ready (why create kubernetes manifests for such a small app? Well, why not? :) )
The necesary manifest are in the Kubernetes folder and you can apply them to your cluster with a simple:

```bash
kubectl apply -f .
```


files/History.txt
This is history!

webPageHandlers_test.go
package main

import (
	"./fileManager"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func Test_healthCheckHandler(t *testing.T) {
	tests := []struct {
		name string
		//args args
	}{
		{name: "Check health check"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			expected := `{"alive": true}`
			req, err := http.NewRequest("GET", "/healthCheck", nil)
			if err != nil {
				t.Fatal(err)
			}
			recorder := httptest.NewRecorder()
			handler := http.HandlerFunc(healthCheckHandler)
			handler.ServeHTTP(recorder, req)

			if status := recorder.Code; status != http.StatusOK {
				t.Errorf("Health Check returned the wrong http code: got %v, want %v", status, http.StatusOK)
			}

			if recorder.Body.String() != expected {
				t.Errorf("Health Check bodey contains %v, want %v", recorder.Body.String(), expected)
			}
		})
	}
}

func Test_viewHandler(t *testing.T) {
	tests := []struct {
		name     string
		title    string
		wantErr  bool
		expected string
	}{
		{name: "View test history page", title: "History", wantErr: false, expected: ""},
		{name: "Invalid page name", title: "xyz", wantErr: true, expected: "<a href=\"/edit/xyz\">Found</a>"},
	}
	fileManager.SaveDisplayTestPage(filesDir, txtExtension, readWriteFileMode)
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			title := tt.title
			req, err := http.NewRequest("GET", "/view/"+title, nil)
			if err != nil {
				t.Fatal(err)
			}
			recorder := httptest.NewRecorder()
			handler := makeHandler(viewHandler)
			handler.ServeHTTP(recorder, req)

			if tt.wantErr {
				if status := recorder.Code; status != http.StatusFound {
					t.Errorf("Attempting to view non existing file returned the wrong http code: got %v, want %v", status, http.StatusFound)
				}
				if !strings.Contains(recorder.Body.String(), "/edit/"+title) {
					t.Errorf("Expected redirection. Wanted address /edit/"+title+", got %v ", recorder.Body.String())
				}
			} else {
				if status := recorder.Code; status != http.StatusOK {
					t.Errorf("Attempting to view "+title+" file returned the wrong http code: got %v, want %v", status, http.StatusOK)
				}

				if recorder.Body.String() == "" {
					t.Errorf("Recieved an empty body %v", recorder.Body.String())
				}
			}

		})
	}
}


fileManager/fileMgr_test.go
package fileManager

import (
	"os"
	"testing"
)

func TestGetFileName(t *testing.T) {
	type args struct {
		title        string
		filesDir     string
		txtExtension string
	}
	tests := []struct {
		name    string
		args    args
		want    string
		wantErr bool
	}{
		{name: "Positive test",
			args:    args{title: "MyTitle", filesDir: "files/", txtExtension: ".txt"},
			want:    "files/MyTitle.txt",
			wantErr: false,
		},
		{name: "No directory provided",
			args:    args{title: "MyTitle", filesDir: "", txtExtension: ".txt"},
			want:    "MyTitle.txt",
			wantErr: false,
		},
		{name: "Missing title",
			args: args{title: "", filesDir: "files/", txtExtension: ".txt"},
			//want: "invalid title or extension",
			wantErr: true,
		},
		{name: "Missing extension",
			args: args{title: "MyTitle", filesDir: "files/", txtExtension: ""},
			//want: "invalid title or extension",
			wantErr: true,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := GetFileName(tt.args.title, tt.args.filesDir, tt.args.txtExtension)
			if (err != nil) != tt.wantErr {
				t.Errorf("GetFileName() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if got != tt.want {
				t.Errorf("GetFileName() got = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestListDirFiles(t *testing.T) {
	type args struct {
		directory  string
		dateFormat string
	}
	tests := []struct {
		name    string
		args    args
		want    FileListing
		wantErr bool
	}{
		{name: "Find History file",
			args: args{directory: "../files/", dateFormat: "Jan-02-06"},
			want: FileListing{
				Count:     0,
				Names:     nil,
				FullNames: nil,
				DirName:   "../files/",
				Files:     []fileSpec{{Name: "History", FullName: "History.txt"}},
			},
			wantErr: false,
		},
		{name: "Invalid file",
			args:    args{directory: "xyz/", dateFormat: "Jan-02-06"},
			wantErr: true,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := ListDirFiles(tt.args.directory, tt.args.dateFormat)
			if (err != nil) != tt.wantErr {
				t.Errorf("ListDirFiles() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if !tt.wantErr {
				foundHistory := false
				for _, file := range got.Files {
					if file.FullName == tt.want.Files[0].FullName && file.Name == tt.want.Files[0].Name {
						foundHistory = true
					}
				}
				if !foundHistory {
					t.Errorf("ListDirFiles() got = %v, missing %s", got, tt.want.Files[0].FullName)
				}
			}
		})
	}
}

func TestPage_Save(t *testing.T) {
	type fields struct {
		Title string
		Body  []byte
	}
	type args struct {
		filesDir     string
		txtExtension string
		fileMode     os.FileMode
	}
	tests := []struct {
		name    string
		fields  fields
		args    args
		wantErr bool
	}{
		{name: "Save History File",
			fields: fields{
				Title: "History",
				Body:  []byte("This is history!"),
			},
			args: args{
				filesDir:     "../files/",
				txtExtension: ".txt",
				fileMode:     0600,
			},
			wantErr: false,
		},
		{name: "Missing file name",
			fields: fields{
				Title: "",
				Body:  []byte("This is history!"),
			},
			args: args{
				filesDir:     "files",
				txtExtension: ".txt",
				fileMode:     0600,
			},
			wantErr: true,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			pg := &Page{
				Title: tt.fields.Title,
				Body:  tt.fields.Body,
			}
			if err := pg.Save(tt.args.filesDir, tt.args.txtExtension, tt.args.fileMode); (err != nil) != tt.wantErr {
				t.Errorf("Save() error = %v, wantErr %v", err, tt.wantErr)
			}
		})
	}
}

func TestSaveDisplayTestPage(t *testing.T) {
	type args struct {
		filesDir     string
		txtExtension string
		fileMode     os.FileMode
	}
	tests := []struct {
		name string
		args args
	}{
		{name: "Save and Display test page",
			args: args{
				filesDir:     "files",
				txtExtension: ".txt",
				fileMode:     0600,
			},
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
		})
	}
}


webPageHandlers.go
package main

import (
	"./fileManager"
	"io"
	"net/http"
)

//makeHandler: Take a http request to save/edit/view a file , validate urk, get file title and call the appropriate function
// Note that we use closures to pass the  function we will need to call at the end of the day as
// the first parameter to this function
func makeHandler(fn func(http.ResponseWriter, *http.Request, string)) http.HandlerFunc {
	return func(responseWriter http.ResponseWriter, r *http.Request) {
		match := validPath.FindStringSubmatch(r.URL.Path)
		if match == nil {
			http.NotFound(responseWriter, r)
			return
		}
		fn(responseWriter, r, match[2])
	}
}

//handler: handle http request
func handler(responseWriter http.ResponseWriter, r *http.Request) {
	_ = r
	files, err := fileManager.ListDirFiles(filesDir, dateFormat)
	err = templates.ExecuteTemplate(responseWriter, indexTemplate, files)
	if err != nil {
		http.Error(responseWriter, err.Error(), http.StatusInternalServerError)
	}
}

//viewHandler: Handle request to view a file
func viewHandler(responseWriter http.ResponseWriter, r *http.Request, title string) {
	//title, err := getTitle(responseWriter, r)
	//if err != nil {
	//	return
	//}
	pg, err := fileManager.LoadPage(title, filesDir, txtExtension)
	if err != nil {
		http.Redirect(responseWriter, r, editUrl+title, http.StatusFound)
		return
	}
	renderTemplate(responseWriter, pg, viewTemplate)

}

//editHandler: Handle request to edit a file
func editHandler(responseWriter http.ResponseWriter, r *http.Request, title string) {
	_ = r // added since we still need the handlers to be uniform so this can be called from the makeHandler function
	pg, err := fileManager.LoadPage(title, filesDir, txtExtension)
	if err != nil {
		pg = &fileManager.Page{Title: title}
	}
	renderTemplate(responseWriter, pg, editTemplate)
}

// saveHandler: Get data from the request body and save it into a new file
func saveHandler(responseWriter http.ResponseWriter, r *http.Request, title string) {
	body := r.FormValue("body")
	pg := fileManager.Page{
		Title: title,
		Body:  []byte(body),
	}
	err := pg.Save(filesDir, txtExtension, readWriteFileMode)
	if err != nil {
		http.Error(responseWriter, err.Error(), http.StatusInternalServerError)
	}
	http.Redirect(responseWriter, r, viewUri+title, http.StatusFound)
}

func newHandler(responseWriter http.ResponseWriter, r *http.Request) {
	match := validFileName.FindStringSubmatch(r.FormValue("fileName"))
	if match == nil {
		http.Error(responseWriter, "Invalid file name. Please use only letters and numbers and do not enter extension.", http.StatusBadRequest)
		return
	}
	title := match[1]
	http.Redirect(responseWriter, r, viewUri+title, http.StatusFound)
}

func healthCheckHandler(responseWriter http.ResponseWriter, r *http.Request) {
	// A very simple health check.
	_ = r
	responseWriter.WriteHeader(http.StatusOK)
	responseWriter.Header().Set("Content-Type", "application/json")

	_, err := io.WriteString(responseWriter, `{"alive": true}`)
	if err != nil {
		http.Error(responseWriter, err.Error(), http.StatusInternalServerError)
	}
}

//renderTemplate: Load html template from the template cache and render page content to send back to client
func renderTemplate(responseWriter http.ResponseWriter, pg *fileManager.Page, templateName string) {
	err := templates.ExecuteTemplate(responseWriter, templateName, pg)
	if err != nil {
		http.Error(responseWriter, err.Error(), http.StatusInternalServerError)
	}
}


fileManager/fileMgr.go
package fileManager

import (
	"errors"
	"fmt"
	"io/ioutil"
	"os"
)

// Page: Holds page we need to save/load
type Page struct {
	Title string
	Body  []byte
}

// fileSpec: File system details for a file
type fileSpec struct {
	Name     string
	ModTime  string
	Size     int64
	FullName string
}

//FileListing: List of filenames in a directory
type FileListing struct {
	Count     int
	Names     []string
	FullNames []string
	DirName   string
	Files     []fileSpec
}

//GetFileName: build file name base on tile + the file extension
func GetFileName(title, filesDir, txtExtension string) (string, error) {
	if title == "" || txtExtension == "" {
		return "", errors.New("invalid title or extension")
	}
	return filesDir + title + txtExtension, nil
}

// loadFileListing: Get the list of files in a given directory
func loadFileListing(dirName string) ([]os.FileInfo, error) {
	fileListing, err := ioutil.ReadDir(dirName)
	if err != nil {
		return nil, err
	}
	return fileListing, nil
}

//ListDirFiles: Load the count and names of the files in a directory into the FileListing struct
func ListDirFiles(directory, dateFormat string) (FileListing, error) {
	files := FileListing{DirName: directory}
	filesInDir, err := loadFileListing(directory)
	if err != nil {
		//Do nothing
	} else {
		fileSpecs := make([]fileSpec, len(filesInDir))
		for i, info := range filesInDir {
			if !info.IsDir() {
				fileSpecs[i] = fileSpec{
					FullName: info.Name(),
					Name:     info.Name()[:len(info.Name())-4],
					ModTime:  info.ModTime().Format(dateFormat),
					Size:     info.Size(),
				}
			}
		}
		files = FileListing{
			Count:   len(fileSpecs),
			DirName: directory,
			Files:   fileSpecs,
		}
	}
	return files, err
}

//LoadPage: load page from disk into a page struct
func LoadPage(title string, filesDir, txtExtension string) (*Page, error) {
	fileName, err := GetFileName(title, filesDir, txtExtension)
	if err != nil {
		return nil, err
	}
	body, err := ioutil.ReadFile(fileName)
	if err != nil {
		return nil, err
	}
	return &Page{title, body}, nil
}

//save: Save a page to disk
func (pg *Page) Save(filesDir, txtExtension string, fileMode os.FileMode) error {
	fileName, err := GetFileName(pg.Title, filesDir, txtExtension)
	if err != nil {
		return err
	}
	return ioutil.WriteFile(fileName, pg.Body, fileMode)

}

//saveDisplayTestPage: Save a string into a file. Reload the file and display content.
func SaveDisplayTestPage(filesDir, txtExtension string, fileMode os.FileMode) {
	pageTitle := "History"
	pg := Page{
		Title: pageTitle,
		Body:  []byte("This is history!"),
	}
	err := pg.Save(filesDir, txtExtension, fileMode)
	if err != nil {
		panic(err)
	}
	pgLd, err := LoadPage(pageTitle, filesDir, txtExtension)
	if err != nil {
		panic(err)
	} else {
		fmt.Println(pgLd.Title + ": " + string(pgLd.Body))
	}
}


webApp.go
package main

import (
	"./fileManager"
	"html/template"
	"log"
	"net/http"
	"regexp"
)

//dateFormat: Formatting to display dates in the screen
const dateFormat string = "Jan-02-06"

// File manipulation constants
const (
	txtExtension      string = ".txt"
	readWriteFileMode        = 0600
	filesDir                 = "files/"
)

// Html templates
const (
	templatesDir  = "htmlTemplates/"
	editTemplate  = "edit.html"
	viewTemplate  = "view.html"
	indexTemplate = "index.html"
)

// Page uris
const (
	editUrl = "/edit/"
	//saveUri = "/save/"
	viewUri = "/view/"
)

//templates: Html templates cache loaded from disk
var templates = template.Must(template.ParseFiles(templatesDir+editTemplate, templatesDir+viewTemplate, templatesDir+indexTemplate))

//validPath: Regular expression to be validate that we have a valid path to save and retrieve files
var validPath = regexp.MustCompile("^/(edit|save|view)/([a-zA-Z0-9]+)$")

//validFileName: Validate that the file name only contains letters and numbers
var validFileName = regexp.MustCompile("^([a-zA-Z0-9]+)$")

//startServer: Start web server and call handlers
func startServer() {
	http.HandleFunc("/", handler)
	http.HandleFunc("/new/", newHandler)
	http.HandleFunc("/view/", makeHandler(viewHandler))
	http.HandleFunc("/edit/", makeHandler(editHandler))
	http.HandleFunc("/save/", makeHandler(saveHandler))
	http.HandleFunc("/healthCheck", healthCheckHandler)
	log.Fatal(http.ListenAndServe(":8080", nil))
}

func main() {
	fileManager.SaveDisplayTestPage(filesDir, txtExtension, readWriteFileMode)
	startServer()
}
